### PR TITLE
Ensure all jobset attributes evaluate, and check that in CT

### LINKED ISSUE #269357: outpath-listing / enumerating release jobset is single-threaded and slow
The only known way to enumerate the names of all packages/jobs in a release is the one used by ofborg for the eval-checks:

- https://github.com/nixos/ofborg#running-meta-checks-locally

This is obscenely slow and requires an absurd amount of memory.  It is entirely single-threaded with zero parallelism.

This accounts for the vast majority of the time for each CI run.

### PR DESCRIPTION
**Pull request by ghost**
#### Description

aka ["eval checks go brrrr"](https://github.com/NixOS/nixpkgs/pull/269403)

Please review the commits individually.  Produced as part of the non-Hydra tooling used to validate

- https://github.com/NixOS/nixpkgs/pull/268097

#### Description of changes

This PR adds `release-attrpaths-superset.nix`, which walks the entire nixpkgs tree and emits a superset of all release attrnames in only 44 seconds on a 3ghz CPU, using less than 5 gbytes of memory.  By comparison, on the same CPU the `nix-env -qaP` hack used by ofborg on every PR submission requires waiting **25 minutes** and peaks at 58 gbytes, even with checkMeta turned off.  Full details below.  This is:

- 33x faster (or 2.9% of the elapsed time)
- 12.5x less memory (or 8.0% of the peak memory usage)

If all you need is the attrnames -- for example to launch a mass rebuild -- then you're done.  If you want to replace the ofborg `outpaths.nix` step, see the follow-up PR:

- https://github.com/NixOS/nixpkgs/pull/269403

#### Validation

To confirm that the `release-attrpaths-superset.nix` really are a superset of the slower method, run

```
nix-build pkgs/test/release/release-attrpaths-superset.nix -A failures
```

#### Additional CT ([Continuous Testing](https://en.wikipedia.org/wiki/Continuous_testing)) check

This PR adds `recurseForRelease=false` where necessary within nixpkgs in order to make this possible -- it screens out infinite recursions and several non-`tryEval`-catchable failures (for example, [passing a mistyped argument to a builtin](https://nixos.org/manual/nix/unstable/language/builtins#builtins-tryEval), which is an uncatchable error) and infinite recursions.

Finding all the places where it was necessary to add this was a lot of tedious work, so to make sure that work is not undone, `lib/tests/release.nix` now verifies that `release-attrpaths-superset.nix` evaluates successfully.

Although it doesn't really fit in `lib/tests`, there isn't anywhere else it can be placed and still be run by ofborg on every commit.  If some other such place materializes in the future I would be happy to relocate it, but since ofborg is out-of-tree this PR should not block on that happening.

#### Fixes

- https://github.com/NixOS/nixpkgs/issues/269357

### CODE REVIEW

**Comment by amjoseph-nixpkgs** [comment]
#### Measurements

According to GNU Time `release-attrpaths-superset.nix` consumes:,

```
Command being timed: "nix-instantiate --eval --strict --json pkgs/top-level/release-attrpaths-superset.nix -A names"
User time (seconds): 44.88
System time (seconds): 8.09
Percent of CPU this job got: 99%
Elapsed (wall clock) time (h:mm:ss or m:ss): 0:53.20
Average shared text size (kbytes): 0
Average unshared data size (kbytes): 0
Average stack size (kbytes): 0
Average total size (kbytes): 0
Maximum resident set size (kbytes): 4823028
Average resident set size (kbytes): 0
```

Compared to `release-outpaths.nix`:

```
Command being timed: "nix-env -qaP --no-name --out-path --arg checkMeta false --argstr path /git/work/pr/release-outpaths -f pkgs/top-level/release-outpaths.nix"
User time (seconds): 1294.48
System time (seconds): 224.37
Percent of CPU this job got: 99%
Elapsed (wall clock) time (h:mm:ss or m:ss): 25:25.01
Average shared text size (kbytes): 0
Average unshared data size (kbytes): 0
Average stack size (kbytes): 0
Average total size (kbytes): 0
Maximum resident set size (kbytes): 58336672
Average resident set size (kbytes): 0
```

**Comment by piegamesde** [comment]
If I understand it correctly, this injects a magic attribute `recurseForRelease` into all package sets, similarly to `recurseIntoAttrs`. I already dislike `recurseIntoAttrs` for many reasons, so I'd like to know if there are other solutions to achieve the same goal. For example, could this be merged with `recurseIntoAttrs` by making the latter more than just a boolean?

**Comment by amjoseph-nixpkgs** [comment]
I too dislike `recurseForAttrs`, particularly because it has magical effects on `nix-env` (i.e. the interpreter).

`recurseForRelease` is temporary; I would like to see it removed after the following events happen:

1. It becomes possible to walk `all-packages.nix` tree without eval failing (this PR)
2. We add CI which prevents people from merging changes that break #1 (also this PR)
3. We encourage package maintainers to fix their broken eval and remove `recurseForRelease={true,false}` flags from their packages.
4. We deprecate `recurseForRelease`
5. We remove `recurseForRelease`

Essentially `recurseForRelease` was a way for me to do #1 and #2 possible without deep invasive changes into other peoples' packages, like totally restructuring them.

The intended semantics of `recurseForRelease` are that `release-attrpaths-superset.nix` walks the attrset tree, starting at `pkgs`, and:
- Stops if it encounters `recurseForRelease=false`
- Stops at any value for which `isDerivation` is `true` *unless* that derivation has `recurseForRelease=true`
  - This is, unfortunately, necessary because there are Hydra release jobnames which have proper prefixes which are attrnames of derivations (wtf).

I will add a comment with the above semantics.

**Comment by piegamesde** [comment]
Thank you. One thing I still don't understand (but probably simply because I don't understand release job sets yet) is, why this would be necessary on package sets which don't have `recurseIntoAttrs`? Why would it recurse into them at all in the first place?

Also if your hope is that Nixpkgs one day fully evaluates without errors, I fear that this might never happen. A more realistic approach would be to have a `catch` function truly worthy of its name. (Or a more lazy package system so that we don't have to drag tons of stuff through `mkDerivation`)

File: pkgs/top-level/release-attrpaths-superset.nix
Comment (infinisil) [inline pkgs/top-level/release-attrpaths-superset.nix:81]:
```suggestion
           ! (value.__attrsFailEvaluation or false) then
```

I suggest this be renamed. Also use `!` to be less confusing

File: pkgs/top-level/release-attrpaths-superset.nix
Comment (infinisil) [inline pkgs/top-level/release-attrpaths-superset.nix:1]: The `releaseForRelease`/`__attrsFailEvaluation` should have a description in here.

File: pkgs/top-level/release-attrpaths-superset.nix
Comment (infinisil) [inline pkgs/top-level/release-attrpaths-superset.nix:84]:
```suggestion
        else if ! (value.recurseForRelease or true) then []
```

File: pkgs/top-level/release-outpaths.nix
Comment (infinisil) [inline pkgs/top-level/release-outpaths.nix:11]: I guess this and the `lib/tests/release.nix` change is only meant to be here until ofborg switches to `release-attrpaths-superset.nix`? Because otherwise this file would be better inlined

**Review by infinisil — CHANGES_REQUESTED** [review]
Having taken a very close look at this, I observe that the difference between `recurseForDerivations` and `recurseForRelease` is that:
- `recurseForDerivations`: Used by `nix-env` to show the list of packages. Used by Hydra (through `release.nix`) to decide which attributes to recurse into for building, defaulting to _false_ if not present. This means packages need to be explicitly added in order to be recursed into and shown to users.
- `recurseForRelease`: Intended to be only used by the new `release-attrpaths-superset.nix`, which defaults to _true_ for attribute sets if not present, therefore mirroring what ofborg would do to check the evaluation. The reason for having this is to explicitly prevent evaluation of code paths that cannot be evaluated (and also not caught using `tryEval`).
  - `recurseForRelease` also existed before this PR, but there's currently no definition of it anymore in master, and only some code that checks for it.


Regarding this PR generally, I think it's really good! There are some minor and some less minor suggestions, but overall I think we should have this

File: pkgs/desktops/gnome/extensions/default.nix
Comment (piegamesde) [inline pkgs/desktops/gnome/extensions/default.nix:59]: I find this diff rather unsatisfying. Would it be possible to move the `//` up into `all-packages.nix` instead?

**Comment by amjoseph-nixpkgs** [comment]
> I will add a comment with the above semantics.

I have done this (commit XXXX) <- will edit when I push it

> is, why this would be necessary on package sets which don't have `recurseIntoAttrs`? Why would it recurse into them at all in the first place?

`recurseIntoAttrs` is magical, and affects `nix-env` in ways that are not expressed in Nix code.  `nix-env` contains an attrset-walking routine written in C++ which checks for `recurseIntoAttrs`.  Its semantics are poorly-defined.

The new `release-attrpaths-superset.nix` contains an attrset-walker which is written entirely in Nix.  It does not use the C++ attrset walker, and therefore totally ignores `recurseIntoAttrs`.

> Also if your hope is that Nixpkgs one day fully evaluates without errors,

I do.  In fact, that day is today.  Go ahead, check out this PR and run the command at the top of `release-attrpaths-superset.nix`.

**Comment by amjoseph-nixpkgs** [comment]
@piegamesde everything in [this comment](https://github.com/NixOS/nixpkgs/pull/269356#pullrequestreview-1751687080) by @infinisil is correct and does a much better job of answering your question than I did.

File: lib/tests/release.nix
Comment (amjoseph-nixpkgs) [reply lib/tests/release.nix:70]: The problem is that ofborg is out of tree and tends to ignore pull requests.

I don't think this PR should block on the ofborg maintainers getting around to reviewing and merging the PR.  But I would definitely be okay with reverting this change whenever they *do* get around to that.

This is a big PR; it's going to turn into merge-conflict hell if I have to wait for them.

File: pkgs/top-level/release-attrpaths-superset.nix
Comment (amjoseph-nixpkgs) [reply pkgs/top-level/release-attrpaths-superset.nix:1]: I want to point out that `pkg.recurseForRelease==true`, `pkg.recurseForRelease==false`, and `!(pkg?recurseForRelease)` do three different things.

I'll split this into `__attrsFailEvaluation` and `__recurseIntoDerivation`.  We can make further changes if that doesn't look right.

File: pkgs/top-level/release-outpaths.nix
Comment (amjoseph-nixpkgs) [reply pkgs/top-level/release-outpaths.nix:11]: Yes!  Exactly.

File: pkgs/top-level/release-attrpaths-superset.nix
Comment (amjoseph-nixpkgs) [reply pkgs/top-level/release-attrpaths-superset.nix:1]: See 924f8a5d4fb3d58d0d1fd2857a00e670e3036257

**Comment by amjoseph-nixpkgs** [comment]
@infinisil thank you for the incredibly detailed review!  I believe I have addressed all of your requests except:

- [ ] https://github.com/NixOS/nixpkgs/pull/269356#discussion_r1406958278 (still working on this)
- [ ] https://github.com/NixOS/nixpkgs/pull/269356#discussion_r1406967782 where I would ask that you reconsider having this PR block on changes to ofborg, since it's out-of-tree and this is a fairly large and fragile-while-unmerged PR.  I think we agree on the end goal, I just want to get there by letting the nixpkgs and ofborg PRs be processed concurrently rather than sequentially.

Also I need to address this comment by @piegamesde
- [ ] https://github.com/NixOS/nixpkgs/pull/269356#discussion_r1409136391

Let me make sure CI still passes and then I will wrap up these last few changes.

File: pkgs/desktops/gnome/extensions/default.nix
Comment (amjoseph-nixpkgs) [reply pkgs/desktops/gnome/extensions/default.nix]: How about 1d5a3536372ef2eec1725adf275d66f12a282f2d?

File: pkgs/development/interpreters/lua-5/default.nix
Comment (amjoseph-nixpkgs) [reply pkgs/development/interpreters/lua-5/default.nix]: I believe this is resolved now in e13d9f7a433e4821219abfcaaea66ab26c355b9d

**Comment by amjoseph-nixpkgs** [comment]
Okay, the latest push squashes and does major cleanup to the version history.

I believe it addresses all of the issues raised, except for the one listed below.  @infinisil @piegamesde would you mind taking one more look at this?

The only unresolved issue is the question of whether to:

1. Block this PR on waiting for changes to ofborg or
2. Temporarily place the `release-attrpaths-superset.nix` test in a suboptimal location (`lib/tests`) until ofborg gets around to adding a non-`lib` test entry point (maybe `pkgs/test/release/default.nix`).  As soon as ofborg does that, we can move the test to its final resting location.

Obviously I think we should go with (2).  But not only because this is my PR.  In general, forcing certain changes to jump through more artificial hoops than other changes is a dangerous path to head down.  Nixpkgs is like wikipedia -- anybody can edit anything, subject only to the review process, CI, and a committer merging it.  On an emotional level nobody really likes other people editing what they've written, but we all have to put that aside in order to produce excellent software.

**If we allow there to exist mechanisms (like ofborg being randomly in another repo with a smaller and less-responsive committer group) by which people can *artificially* make it more difficult to change certain pieces of code, there will be a creeping temptation for people to apply those mechanisms to their own changes.  In effect, a way of making their own work "more permanent" than other peoples'.  This is really bad, and we should resist it on principle.**

I think it's bonkers that ofborg isn't in the monorepo, but I accept that that isn't going to change anytime soon.  Until it does change, I feel very strongly that ofborg's separate-repo-ness must never be an additional impediment, no matter how small, to any change being made.

**Comment by Mic92** [comment]
Second round of cherry-picks: https://github.com/NixOS/nixpkgs/pull/271123

**Comment by Mic92** [comment]
Having the actual code that ofborg uses in nixpkgs certainly makes it easier to reproduce evaluation errors locally. My plan is to use `pkgs/top-level/release-outpaths.nix` in nixpkgs-review soon.

**Comment by infinisil** [comment]
Oh there is something else to consider here: ofborg also fails if the evaluation outputs anything on stderr (aka `builtins.trace`). This is generally a good thing, since it prevents Nixpkgs from using deprecated functions (that use e.g. `lib.warn`).

We should have the same thing with the new CI check introduced here.

**Comment by amjoseph-nixpkgs** [comment]
Rebased.

**Comment by amjoseph-nixpkgs** [comment]
> Then how about using GitHub Actions to test this instead for now?

So, there are really *two* problems with ofborg:

1. You can't run it locally
2. It's out-of-tree

Your idea solves the second problem, so it's definitely an improvement, but we still have the first problem.

Could we have one, single "run all the CI checks" entrypoint from githubland into nixpkgs, and then fan out to the various checks using Nix code?  Then we could add the github magic for that single entry point once and never touch it again.  As long as it's a single command it'll be locally reproducible.

I just don't want to see more and more weird YAML machinery accumulating.  We should keep all the logic in Nix, not github-YAML.

**Comment by amjoseph-nixpkgs** [comment]
> We should have the same behavior with the new CI check introduced here.

Okay, see latest push, which adds `enableWarnings?true`; it can be set to `false` for the CI run.

I'd like to leave the default as `true` in order to nag people to work on fixing the eval failures that are currently being dealt with by `tryEval`.  There are only 153 of them and most of those come from big blocks with a common cause (cough cough freebsd cough cough).

If that number ever drops to zero then we can switch on `enableWarnings=true` in CI and it will *stay* at zero (hooray!) and we can even remove the `builtins.tryEval`!

**Review by infinisil — DISMISSED** [review]
(just a 5 minute review)

**Comment by amjoseph-nixpkgs** [comment]
@infinisil please see in particular the last two commits.

File: pkgs/top-level/release-outpaths.nix
Comment (Mic92) [inline pkgs/top-level/release-outpaths.nix:3]: How do you take advantage of parallelization in this code using the information from above? Or is this still to be done? I am asking because of nixpkgs-review.

Comment (Mic92) [reply pkgs/top-level/release-outpaths.nix:3]: Never mind just saw: https://github.com/NixOS/nixpkgs/pull/269403

**Comment by bjornfor** [comment]
>Ensure all jobset attributes evaluate, and check that in CT

Do you mean "CI" here?

**Comment by Mic92** [comment]
> > Ensure all jobset attributes evaluate, and check that in CT
> 
> Do you mean "CI" here?

https://en.wikipedia.org/wiki/Continuous_testing is indeed a thing but also I didn't know what ct was standing for until I saw it written out in the test file. For clarity we should probably stick to ci as it's way more known abbreviation.

**Comment by amjoseph-nixpkgs** [comment]
> https://en.wikipedia.org/wiki/Continuous_testing is indeed a thing

It always has been :smile:

> For clarity we should probably stick to

Alright, fixing the acronym can wait until another day.  I've pushed 1f8232eb1e35b64ec2562cae2c147164e74652d0 which changes `<nixpkgs/ct>` to `<nixpkgs/tests>`.

**Comment by infinisil** [comment]
I'd think it should be `<nixpkgs/ci>`. Seeing a toplevel `tests` would make me think all of Nixpkgs tests are in there.

File: tests/tasks.nix
Comment (infinisil) [inline tests/tasks.nix:59]: Why generate a makefile? Couldn't we just generate a shell script? Let's keep it simple where possible

File: tests/default.nix
Comment (infinisil) [inline tests/default.nix:2]: Not a fan of this. How about a simple `nix-build` that builds a script?

File: pkgs/top-level/release-attrpaths-superset.nix
Comment (infinisil) [inline pkgs/top-level/release-attrpaths-superset.nix:1]: With `<nixpkgs/[tests|ct|ci]>` existing, I think this file would fit better in there. Same for `release-outpaths.nix`.

**Review by infinisil — DISMISSED** [review]

**Comment by infinisil** [comment]
Btw this PR has gotten bigger in scope now (also adding a general CI task specification), which makes reviewing and merging this even harder. Generally I'd really prefer to keep PRs as small as possible.

File: tests/tasks.nix
Comment (roberth) [reply tests/tasks.nix:59]: I agree with @infinisil, and logic like this, if we need it should go in its own function (`lib.generators`?) and have tests.
Non-trivial test support code needs tests, just like any other code.

**Comment by amjoseph-nixpkgs** [comment]
> Seeing a toplevel `tests` would make me think all of Nixpkgs tests are in there.

That's the point: all tests should be accessed via that entry point, and only that entry point.

Of course tests can pull in things that aren't under that directory, just like `lib/tests` and `pkgs/test` already pull in things that aren't under those directories.

File: tests/tasks.nix
Comment (amjoseph-nixpkgs) [reply tests/tasks.nix:59]: > Why generate a makefile?

Parallelism.

`make` is the simplest tool that knows how to run tasks with dependencies in parallel.

Our ofborg-driven CT is already parallel, so if the single-entry-point tests are run serially any attempt to migrate the existing tests to the new entry point will cause a major usability regression.

Comment (infinisil) [reply tests/tasks.nix:59]: For now we don't need this though, there's only a single task. And the one you commented out for the future doesn't depend on it either. In fact, both of them are just derivations, so we can just use Nix itself to run them in parallel! The comment about these tasks running "outside of nix-build" doesn't seem to apply.

File: tests/default.nix
Comment (amjoseph-nixpkgs) [reply tests/default.nix:2]: I have rolled back to before the point where the top-level entry point was added.  I've also linked this comment here:

- https://github.com/NixOS/nixpkgs/issues/272591

so we don't lose this feedback.

**Review by infinisil — APPROVED** [review]
There's some minor things I could complain about, but overall this is fine!

Note that immediately this does just increase the load on ofborg, but it's not by much, and this enables further follow-up PR's, so I'd say let's go for it.

I can merge once you rebase!

**Comment by flokli** [comment]
Needs a rebase.

**Comment by amjoseph-nixpkgs** [comment]
Rebased.

**Review by infinisil — APPROVED** [review]
Looking good to me, tested locally. To prevent further conflicts I'll merge now! :rocket:

Edit: Hold on, I'll also quickly test this on Darwin

**Pull request merged**
